File: source/PointOfSalesV2.Entities/Product.cs

global using static PointOfSalesV2.Common.Enums;

namespace PointOfSalesV2.Entities
{
    public class Product : CommonData
    {

        [MaxLength(200)]
        [Translate]
        [Export(Order = 1)]
        public string Description { get; set; }
        [Export(Order = 6)]
        public decimal Price { get; set; }
        [Export(Order = 7)]
        public decimal Price2 { get; set; }
        [Export(Order = 8)]
        public decimal Price3 { get; set; }
        [MaxLength(50)]
        [Export(Order = 3)]
        public string Code { get; set; }
        [NotMapped]
        public long OldId { get; set; }
        public char Type { get; set; } = 'C';

        [NotMapped]
        public AppointmentTypes ProductType 
        {
            get 
            {
                return (AppointmentTypes)this.Type;
            }
        }

        public long? MedicalSpecialityId { get; set; }
        public string DetailsClass { get; set; }
        public bool IsService { get; set; }
        public bool IsCompositeProduct { get; set; }
        [Export(Order = 5)]
        public decimal Cost { get; set; }
        [Export(Order = 9)]
        public decimal SellerRate { get; set; } = 0;
        
        public long CurrencyId { get; set; }
        public decimal Existence { get; set; }
        [MaxLength(100)]
        [Translate]
        [Export(Order = 0)]
        public string Name { get; set; }

        [MaxLength(50)]
        [Export(Order = 2)]
        public string Sequence { get; set; }

        public string Details { get; set; }





        [ForeignKey("CurrencyId")]
        [Export(Order = 4, ChildProperty = "Code")]
        public Currency Currency { get; set; }


        [ForeignKey("MedicalSpecialityId")]
        [Export(Order = 4, ChildProperty = "Name")]
        public MedicalSpeciality MedicalSpeciality { get; set; }




        public IEnumerable<ProductTax> Taxes { get; set; }






        public IEnumerable<UnitProductEquivalence> ProductUnits { get; set; }
        public IEnumerable<ProductSupplierCost> SuppliersCosts { get; set; }


        public IEnumerable<CompositeProduct> BaseCompositeProducts { get; set; }



    }
}

File: source/PointOfSalesV2.FrontEnd/ClientApp/src/app/@core/data/product.ts
import { BaseModel } from './baseModel';
import { Currency } from './currencyModel';
import { Tax } from './taxModel';
import { Unit } from './unitModel';


export class Product extends BaseModel {

    description:string ='';
    price:number =0;
    price2:number =0;
    price3:number =0;
    code:string ='';
    isService:boolean = false;
    isCompositeProduct:boolean = false;
    cost:number =0;
    sellerRate:number =0;
    currencyId:number =0;
    existence:number =0;
    name:string ='';
    details:string ='';
    currency :Currency | null = null
    medicalSpecialityId:number=0;
    taxes:any[]=[];
    type:string ='';
    medicalSpeciality:any;
    selected:boolean=false;
    productUnits:any[]=[];
    baseCompositeProducts:any[]=[];
    suppliersCosts:any[]=[];
}
export class ProductTax extends BaseModel {

    productId:number =0;
    taxId:number =0;
    tax:Tax | null = null;
}

export class UnitProductEquivalence extends BaseModel {

         unitId:number =0;
         productId:number =0;
         equivalence:number =0;
         isPrimary :boolean = false;
         order :number =0;
         costPrice :number =0;
         sellingPrice :number =0;
         unit:Unit | null = null;
}

export class CompositeProduct extends BaseModel {

        ProductId :number =0;
        currencyId:number =0;
        BaseProductId:number =0;
        BaseProductUnitId :number =0;
        Quantity :number =0;
        TotalCost :number =0;
        TotalPrice :number =0;
        BaseProduct:Product | null = null;
        Product:Product | null = null;
        UnitProductEquivalence: UnitProductEquivalence | null = null;
        Currency: Currency | null = null;
}
